src/cdk/observers/private/shared-resize-observer.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {inject, Injectable, NgZone, OnDestroy} from '@angular/core';
+import {Observable, Subject} from 'rxjs';
+import {filter, shareReplay, takeUntil} from 'rxjs/operators';
+import {Platform} from '@angular/cdk/platform';
+
+/**
+ * Handler that logs "ResizeObserver loop limit exceeded" errors.
+ * These errors are not shown in the Chrome console, so we log them to ensure developers are aware.
+ * @param e The error
+ */
+const loopLimitExceededErrorHandler = (e: unknown) => {
+  if (e instanceof Error && e.message === 'ResizeObserver loop limit exceeded') {
+    console.error(
+      `${e.message}. This could indicate a performance issue with your app. See https://github.com/WICG/resize-observer/blob/master/explainer.md#error-handling`,
+    );
+  }
+};
+
+/**
+ * A shared ResizeObserver to be used for a particular box type (content-box, border-box, or
+ * device-pixel-content-box)
+ */
+class SingleBoxSharedResizeObserver {
+  /** Stream that emits when the shared observer is destroyed. */
+  private _destroyed = new Subject<void>();
+  /** Stream of all events from the ResizeObserver. */
+  private _resizeSubject = new Subject<ResizeObserverEntry[]>();
+  /** ResizeObserver used to observe element resize events. */
+  private _resizeObserver?: ResizeObserver;
+  /** A map of elements to streams of their resize events. */
+  private _elementObservables = new Map<Element, Observable<ResizeObserverEntry[]>>();
+
+  constructor(
+    /** The box type to observe for resizes. */
+    private _box: ResizeObserverBoxOptions,
+    /** The platform. */
+    platform: Platform,
+  ) {
+    if (platform.isBrowser) {
```

Comment:
I think that we should do `typeof ResizeObserver !== 'undefined'` instead of checking for the server which would cover browsers that don't support `ResizeObserver`.

Reply:
I think al the browsers we officially support also support `ResizeObserver`, but I also don't see any reason not to change the check, so done!